### PR TITLE
Remove item command.

### LINKED ISSUE #6: As a user, I want to delete my tracked items I have registered in the system, so that I can prevent clutter.
D for item class.

### PR DESCRIPTION
**Pull request by blackonyyx**
resolves #6 
Some limited level of Unittesting applied.

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/DeleteItemCommand.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    public static final String COMMAND_WORD = "deli";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Deletes a item in the item list. "
+            + "Parameters: "
+            + PREFIX_ITEM_NAME + "ITEM NAME\n"
+            + "Example: " + COMMAND_WORD + " "
+            + PREFIX_ITEM_NAME + "Iron ";
+    public static final String MESSAGE_SUCCESS = "Item and Recipes of the item Deleted: %1$s";
+    public static final String MESSAGE_ITEM_NOT_FOUND = "Item is not found in the item list";
+    public static final String MESSAGE_ITEM_ALREADY_DELETED = "Item was already previously deleted";
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/commands/DeleteItemCommand.java]: Should this use the NOT_FOUND message instead? User does not need to know our implementation

File: src/main/java/seedu/address/logic/commands/DeleteItemCommand.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        // filter to only get matching and not deleted recipes
+        itemList.removeIf(x -> !x.getName().equals(productName) || x.isDeleted());
+        if (itemList.isEmpty()) {
+            throw new CommandException(MESSAGE_ITEM_NOT_FOUND);
+        }
+        Item itemToDelete;
+        itemToDelete = itemList.stream()
+                .filter(x -> !x.isDeleted())
+                .findFirst() // Get the first (and only) item matching or else throw Error
+                .orElseThrow(()-> new CommandException(MESSAGE_ITEM_ALREADY_DELETED));
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/commands/DeleteItemCommand.java]:
```
itemList.removeIf(x -> !x.getName().equals(productName) || x.isDeleted());
```
in line 43 looks like it already ensures the item won't be deleted

File: src/main/java/seedu/address/logic/commands/DeleteItemCommand.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        model.deleteItem(itemToDelete);
+        List<Recipe> recipeList = new ArrayList<>(model.getFilteredRecipeList());
+        // remove recipes from consideration that are not deleted, are not the product of a recipe, nor contribute
+        // to a recipe.
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/commands/DeleteItemCommand.java]: Suggest to change phrasing to something like
"soft deletes recipes that contain deleted item as product or ingredient"

File: src/main/java/seedu/address/logic/parser/DeleteItemCommandParser.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.util.stream.Stream;
+
+import seedu.address.logic.commands.DeleteItemCommand;
+import seedu.address.logic.parser.exceptions.ParseException;
+
+public class DeleteItemCommandParser implements Parser<DeleteItemCommand> {
+    /**
+     * Parses the given {@code String} of arguments in the context of the DeleteRecipeCommand
+     * and returns a DeleteRecipeCommand object for execution.
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/parser/DeleteItemCommandParser.java]: DeleteRecipeCommand -> DeleteItemCommand

File: src/main/java/seedu/address/model/item/UniqueItemList.java [hunk 5]
```diff
@@ -80,6 +80,21 @@ public void remove(Item toRemove) {
         }
     }
 
+    /**
+     * Deletes the equivilent item from the list, sets deleted flag to true
+     * The item must exist in the list
+     * @param toDelete item to delete
```
Comment (aqcd) [inline src/main/java/seedu/address/model/item/UniqueItemList.java]: use proper punctuation please

File: src/main/java/seedu/address/model/recipe/Ingredient.java [hunk 6]
```diff
@@ -15,5 +15,18 @@
      */
     public Ingredient(Integer key, Integer value) {
         super(key, value);
+        assert value > 0 : "Value should never be 0 or negative";
+    }
+    /**
+     * Utilty function to check if itemid is equivelent to contained id
+     * @param id identity key
+     * @return boolean check that the id is the item
```
Comment (aqcd) [inline src/main/java/seedu/address/model/recipe/Ingredient.java]: use proper punctuation please

**Review by aqcd — DISMISSED** [review]
Tests need to be added back.

File: src/main/java/seedu/address/model/item/UniqueItemList.java (hunk 5 above)
Comment (blackonyyx) [reply src/main/java/seedu/address/model/item/UniqueItemList.java]: Oof

**Comment by blackonyyx** [comment]
From commit 9b2ccce:

In process of Unittesting, dificulty was encountered from utilising
testing utilities of original code base.

Specifically the assertCommandSuccess testutil method, relies on the .equals
of model manager, and model manager equality has null values initialised for AB3
and Inventoryinator versions, which if implemented to check, cause testcase
failures in stubs.

Recommended to perform refactor to code base and remove ab3 artifacts, for more effective
testing.

Will open new issue based on this finding.

File: docs/UserGuide.md [hunk 7]
```diff
@@ -169,19 +169,17 @@ NOT Supported as of v1.1 yet
...
 **NAME:**
-- `del` - delete an item
+- `deli` - delete an item
 
 **SYNOPSIS:**
-- `del` \<item name\> \[-r index\]
+- `deli` \<item name\>
 
 **DESCRIPTION:**
 - **deletes** the item in the inventory with the corresponding item name
```
Comment (aqcd) [inline docs/UserGuide.md]: and all recipes associated with the item

File: src/main/java/seedu/address/logic/commands/DeleteItemCommand.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    public static final String COMMAND_WORD = "deli";
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Deletes a item in the item list. "
+            + "Parameters: "
+            + PREFIX_ITEM_NAME + "ITEM NAME\n"
+            + "Example: " + COMMAND_WORD + " "
+            + PREFIX_ITEM_NAME + "Iron ";
+    public static final String MESSAGE_SUCCESS = "Item and Recipes of the item Deleted: %1$s";
+    public static final String MESSAGE_ITEM_NOT_FOUND = "Item is not found in the item list";
+    private final String productName;
+    /**
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/commands/DeleteItemCommand.java]: please space the methods in this file

File: src/main/java/seedu/address/logic/commands/DeleteRecipeCommand.java [hunk 9]
```diff
@@ -31,7 +31,6 @@
 
     private final String productName;
     private final Index index;
-
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/commands/DeleteRecipeCommand.java]: spacing

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 10]
```diff
@@ -32,7 +32,6 @@
 public class EditCommand extends Command {
 
     public static final String COMMAND_WORD = "edit";
-
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: spacing

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 11]
```diff
@@ -52,7 +51,6 @@
 
     private final Index index;
     private final EditPersonDescriptor editPersonDescriptor;
-
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: spacing

File: src/main/java/seedu/address/logic/parser/DeleteItemCommandParser.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+import seedu.address.logic.parser.exceptions.ParseException;
+
+public class DeleteItemCommandParser implements Parser<DeleteItemCommand> {
+    /**
+     * Parses the given {@code String} of arguments in the context of the DeleteItemCommand
+     * and returns a DeleteItemCommand object for execution.
+     *
+     * @throws ParseException if the user input does not conform to the expected format
+     */
+
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/parser/DeleteItemCommandParser.java]: remove this line

File: src/main/java/seedu/address/model/ItemList.java [hunk 13]
```diff
@@ -81,10 +81,17 @@ public void addItem(Item p) {
      */
     public void setItem(Item target, Item editedItem) {
         requireNonNull(editedItem);
-
         items.setItem(target, editedItem);
     }
 
+    /**
+     * Deletes {@code item} from this {@code ItemList}.
+     * {@code recipe} must exist in the item list.
```
Comment (aqcd) [inline src/main/java/seedu/address/model/ItemList.java]: @code item

File: src/main/java/seedu/address/model/item/Item.java [hunk 14]
```diff
@@ -10,7 +10,6 @@
 import java.util.Set;
 
 import seedu.address.model.tag.Tag;
-
```
Comment (aqcd) [inline src/main/java/seedu/address/model/item/Item.java]: spacing

File: src/main/java/seedu/address/model/item/Item.java [hunk 15]
```diff
@@ -88,14 +87,14 @@ public boolean isDeleted() {
     }
 
     /**
-     * Adds recipe with given id to the item.
+     * Adds Recipe Id to item ids
+     * @param recipeId recipe id connected to this item
```
Comment (aqcd) [inline src/main/java/seedu/address/model/item/Item.java]: proper punctuation please

File: src/main/java/seedu/address/model/item/Item.java [hunk 16]
```diff
@@ -88,14 +87,14 @@ public boolean isDeleted() {
...
      */
     public void addRecipeId(int recipeId) {
         if (!recipeIds.contains(recipeId)) {
             recipeIds.add(recipeId);
         }
     }
-
```
Comment (aqcd) [inline src/main/java/seedu/address/model/item/Item.java]: spacing

File: src/main/java/seedu/address/model/item/Item.java [hunk 17]
```diff
@@ -104,10 +103,24 @@ public boolean isSameItem(Item otherItem) {
         if (otherItem == this) {
             return true;
         }
-
```
Comment (aqcd) [inline src/main/java/seedu/address/model/item/Item.java:107]: spacing

File: src/main/java/seedu/address/model/item/Item.java [hunk 18]
```diff
@@ -104,10 +103,24 @@ public boolean isSameItem(Item otherItem) {
...
         return otherItem != null
                 && otherItem.getName().equals(getName());
     }
+    /**
```
Comment (aqcd) [inline src/main/java/seedu/address/model/item/Item.java]: spacing

File: src/main/java/seedu/address/model/item/Item.java [hunk 19]
```diff
@@ -104,10 +103,24 @@ public boolean isSameItem(Item otherItem) {
...
+     * Returns true if both items have same name, and can be replaced
+     * @param otherItem Other item to compare to this item
+     * @return true if the item of this name can be replaced.
+     */
+    public boolean isReplacable(Item otherItem) {
+        return isSameItem(otherItem) && this.isDeleted();
+    }
+    /**
```
Comment (aqcd) [inline src/main/java/seedu/address/model/item/Item.java]: spacing

File: src/main/java/seedu/address/model/recipe/Ingredient.java [hunk 20]
```diff
@@ -8,12 +8,25 @@
...
+     * Creates a new ingredient.
      *
      * @param key   The key for this pair
      * @param value The value to use for this pair
      */
     public Ingredient(Integer key, Integer value) {
         super(key, value);
+        assert value > 0 : "Value should never be 0 or negative";
+    }
+    /**
```
Comment (aqcd) [inline src/main/java/seedu/address/model/recipe/Ingredient.java]: spacing

File: src/main/java/seedu/address/model/recipe/Ingredient.java [hunk 21]
```diff
@@ -8,12 +8,25 @@
...
+     * Utilty function to check if itemId is equivalent to contained id.
+     * @param id identity key
+     * @return boolean check that the id is the item
+     */
+    public boolean isItem(int id) {
+        return getKey().equals(id);
+    }
+    @Override
```
Comment (aqcd) [inline src/main/java/seedu/address/model/recipe/Ingredient.java]: spacing

File: src/test/java/seedu/address/logic/commands/DeleteItemCommandTest.java [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+    //@Test
+    //public void execute_validIndexUnfilteredList_success() {
+    //    Item itemToDelete = APPLE;
+    //    DeleteItemCommand deleteItemCommand = new DeleteItemCommand(itemToDelete.getName());
+    //    String expectedMessage = String.format(DeleteItemCommand.MESSAGE_SUCCESS, itemToDelete);
+    //    ModelManager expectedModel = new ModelManager(model.getItemList(), model.getLocationList(),
+    //            model.getRecipeList(), new UserPrefs());
+    //    expectedModel.deleteItem(itemToDelete);
+    //    assertCommandSuccess(deleteItemCommand, model, expectedMessage, expectedModel);
+    //}
```
Comment (aqcd) [inline src/test/java/seedu/address/logic/commands/DeleteItemCommandTest.java]: use block comment

File: src/test/java/seedu/address/logic/commands/DeleteItemCommandTest.java [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+    @Test
```
Comment (aqcd) [inline src/test/java/seedu/address/logic/commands/DeleteItemCommandTest.java:34]: fix spacing in this file

File: src/test/java/seedu/address/logic/parser/DeleteItemCommandParserTest.java [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+import static seedu.address.logic.parser.CommandParserTestUtil.assertParseSuccess;
+import static seedu.address.testutil.TypicalItems.APPLE;
+
+import org.junit.jupiter.api.Test;
+
+import seedu.address.logic.commands.DeleteItemCommand;
+
+public class DeleteItemCommandParserTest {
+    private final DeleteItemCommandParser parser = new DeleteItemCommandParser();
+    @Test
```
Comment (aqcd) [inline src/test/java/seedu/address/logic/parser/DeleteItemCommandParserTest.java]: fix spacing in this file

File: src/test/java/seedu/address/model/ItemListTest.java [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.item.Item;
+
+public class ItemListTest {
+    private final ItemList itemList = new ItemList();
+    @Test
```
Comment (aqcd) [inline src/test/java/seedu/address/model/ItemListTest.java]: fix spacing in this file

File: src/test/java/seedu/address/model/item/ItemTest.java [hunk 26]
```diff
@@ -24,6 +24,26 @@ public void asObservableList_modifyList_throwsUnsupportedOperationException() {
         Item item = new ItemBuilder().build();
         assertThrows(UnsupportedOperationException.class, () -> item.getTags().remove(0));
     }
+    /**
```
Comment (aqcd) [inline src/test/java/seedu/address/model/item/ItemTest.java]: spacing

File: src/test/java/seedu/address/model/item/ItemTest.java [hunk 27]
```diff
@@ -24,6 +24,26 @@ public void asObservableList_modifyList_throwsUnsupportedOperationException() {
...
+        Item editedApple = new ItemBuilder(APPLE).build().delete();
+        // same object and marked deleted -> returns true
+        assertTrue(editedApple.isReplacable(APPLE));
+        editedApple = new ItemBuilder(APPLE).withQuantity(VALID_ITEM_QUANTITY_BANANA)
+                .withDescription(VALID_ITEM_DESCRIPTION_BANANA).build();
+        //same object name, but different fields, not deleted -> returns false
+        assertFalse(editedApple.isReplacable(APPLE));
+        editedApple = editedApple.delete();
+        // same object name, but different fields -> returns true
+        assertTrue(editedApple.isReplacable(APPLE));
```
Comment (aqcd) [inline src/test/java/seedu/address/model/item/ItemTest.java]: space out these tests

File: src/test/java/seedu/address/model/item/UniqueItemListTest.java [hunk 28]
```diff
@@ -22,14 +22,12 @@ public void contains() {
         assertTrue(uList.contains(APPLE));
         assertFalse(uList.contains(BANANA));
     }
-
```
Comment (aqcd) [inline src/test/java/seedu/address/model/item/UniqueItemListTest.java:25]: re-add spaces in this file

File: src/test/java/seedu/address/model/item/UniqueItemListTest.java [hunk 29]
```diff
@@ -59,15 +55,26 @@ public void setItem_throwsDuplicateItemException() {
...
+        uList.delete(APPLE); // Item is still in the list, but boolean delete flag is set to true
+        assertTrue(uList.contains(APPLE));
+    }
+    @Test
+    public void delete_throwsItemNotFoundException() {
+        UniqueItemList uList = new UniqueItemList();
+        uList.add(APPLE);
+        assertThrows(ItemNotFoundException.class, ()-> uList.delete(BANANA));
+    }
     @Test
```
Comment (aqcd) [inline src/test/java/seedu/address/model/item/UniqueItemListTest.java]: fix spacing

File: src/test/java/seedu/address/model/recipe/IngredientTest.java [hunk 30]
```diff
@@ -0,0 +1,31 @@
...
+    public void constructor_null_throwsNullPointerException() {
+        assertThrows(NullPointerException.class, () -> new Ingredient(null, null));
+    }
+    @Test
+    public void constructor_invalidIngredient_throwsAssertionException() {
+        Integer x = 1;
+        Integer y = -1;
+        assertThrows(AssertionError.class, () -> new Ingredient(x, y));
+    }
+    @Test
```
Comment (aqcd) [inline src/test/java/seedu/address/model/recipe/IngredientTest.java]: fix spacing

**Review by aqcd — COMMENTED** [review]
Please take note of whitespace usage.

File: src/main/java/seedu/address/logic/commands/DeleteItemCommand.java [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+     * Creates an DeleteItemCommand to delete the specified {@code Item} and connecting recipes
+     * @param itemName The item name to be deleted
+     */
+    public DeleteItemCommand(String itemName) {
+        requireNonNull(itemName);
+        this.productName = itemName;
+    }
+    @Override
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/commands/DeleteItemCommand.java]: spacing

File: src/main/java/seedu/address/logic/commands/DeleteItemCommand.java [hunk 32]
```diff
@@ -0,0 +1,71 @@
...
+                .stream()
+                .noneMatch(z -> z.isItem(itemToDelete.getId())));
+        // delete recipes connected to this identified item as a product, or an ingredient
+        recipeList.forEach(model::deleteRecipe);
+        return new CommandResult(String.format(MESSAGE_SUCCESS, itemToDelete));
+    }
+    public String getProductName() {
+        return productName;
+    }
+    @Override
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/commands/DeleteItemCommand.java]: spacing

File: src/main/java/seedu/address/model/item/UniqueItemList.java [hunk 33]
```diff
@@ -31,12 +30,12 @@
 
     /**
      * Returns true if the list contains an equivalent item as the given argument.
+     * @param toCheck The item to check against.
      */
     public boolean contains(Item toCheck) {
         requireNonNull(toCheck);
         return internalList.stream().anyMatch(toCheck::isSameItem);
     }
-
```
Comment (aqcd) [inline src/main/java/seedu/address/model/item/UniqueItemList.java]: spacing

File: src/main/java/seedu/address/model/recipe/Ingredient.java [hunk 34]
```diff
@@ -8,12 +8,27 @@
...
      */
     public Ingredient(Integer key, Integer value) {
         super(key, value);
+        assert value > 0 : "Value should never be 0 or negative";
+    }
+
+    /**
+     * Utilty function to check if itemId is equivalent to contained id.
+     * @param id identity key
+     * @return boolean check that the id is the item
```
Comment (aqcd) [inline src/main/java/seedu/address/model/recipe/Ingredient.java]: more clear explanation needed

File: src/test/java/seedu/address/model/item/UniqueItemListTest.java [hunk 35]
```diff
@@ -57,17 +63,32 @@ public void setItem_throwsDuplicateItemException() {
...
+    }
+
     @Test
     public void remove() {
         UniqueItemList uList = new UniqueItemList();
         uList.add(APPLE);
         uList.remove(APPLE);
         assertFalse(uList.contains(APPLE));
     }
-
```
Comment (aqcd) [inline src/test/java/seedu/address/model/item/UniqueItemListTest.java:70]: spacing

File: src/test/java/seedu/address/model/item/UniqueItemListTest.java [hunk 36]
```diff
@@ -109,13 +131,14 @@ public void setItems_throwsDuplicateItemException() {
         ArrayList<Item> arr = new ArrayList<>();
         arr.add(APPLE);
         arr.add(APPLE);
+
         assertThrows(DuplicateItemException.class, () -> uList.setItems(arr));
     }
-
```
Comment (aqcd) [inline src/test/java/seedu/address/model/item/UniqueItemListTest.java:114]: spacing

**Review by aqcd — APPROVED** [review]
LGTM after fixing minor cosmetics

**Pull request merged**